### PR TITLE
Add a modified Cholesky decomposition

### LINKED ISSUE #25785: [Feature request] modified Cholesky decomposition
It would be useful to have a modified Cholesky decomposition implemented in PyTorch.

My covariance matrices are singular and `torch.cholesky` fails with RunTimeException
My work-around is to add a small diagonal matrix, but that adds an extra hyper-parameter.

I found one implementation here:
http://faculty.washington.edu/mforbes/doc/pymmf/_build/_generated/api/mmf.math.linalg.cholesky.cholesky_.html

### PR DESCRIPTION
**Pull request by mattip**
Fixes #25785 by adding a modified Cholesky decomposition.

The idea is to add a "small" constant `e`  to the diagonal of the input so that instead of solving for `L` in  `A = L @ L.transpose(-2, -1)`, we solve for `A + (D * e) = L @ L.transpose(-2, -1)` where `D = torch.eye(A.shape[-1])`. The function call becomes `L, e = torch.cholesky_mod(A, upper=True)`. 

The problems I see in this implementation so far are:
- The values of `e` are hard-coded: starting at `e = 1e-8` and increasing by `e *= 100` for 8 tries. Should the value be based on some multiple of the largest eigenvector?
- The code runs sequentially, even on GPU. I could use `magmaCholeskyBatched` but then would need to somehow weed out and "rebatch" the matrices that fail with that value of `e` for the next iteration.
- There are places in the code where I used constructs like `at::fill_` (to assign to a scalar) and `Tensor.copy_`. These do the job, but is there a better way?

I did not touch the problem pointed out in #34272 where raising an error when `info != 0` is slow. One approach would be to return a value in `e` that indicates a failure, maybe negative values?

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -571,6 +571,100 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
...
+    selfTag = self.flatten(0, -3);
+  }
+  
+  scalar_t* self_data = selfTag.data_ptr<scalar_t>();
+  scalar_t* out_data = outTag.data_ptr<scalar_t>();
+  /* TODO: do this in parallel */
+  for (int64_t i = 0; i < batch_size; i++) {
+    scalar_t* out_working_ptr = &out_data[i * self_matrix_stride];
+    scalar_t* self_working_ptr = &self_data[i * self_matrix_stride];
+    double err_val = 1e-8;
```
Comment (vishwakftw) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:601]: Perhaps it would be better to call this `jitter`.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -571,6 +571,100 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
...
+    int info;
+    lapackCholesky<scalar_t>(uplo, n, out_working_ptr, n, &info);
+    for (int64_t j=0; j<8; j++)
+    {
+      if (info == 0) {
+        break;
+      }
+      // lapackChelesky failed. Copy current 2D matrix to out
```
Comment (vishwakftw) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:609]: nit: `lapackCholesky`

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -571,6 +571,100 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
...
+#else
+  char uplo = upper ? 'U' : 'L';
+
+  Tensor outTag =  out;
+  Tensor selfTag = self;
+
+  double* err_data = err.data_ptr<double>();
+  auto self_matrix_stride = matrixStride(self);
+  auto batch_size = batchCount(self);
+  auto n = self.size(-2);
```
Comment (vishwakftw) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:587]: `n == m` for `cholesky` so you can discard one of them.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -571,6 +571,100 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
...
 }
 
+template<typename scalar_t>
+static void  apply_cholesky_mod(const Tensor& self, Tensor & out, Tensor & err, bool upper, std::vector<int64_t>& infos) {
+#ifndef USE_LAPACK
+  AT_ERROR("cholesky: LAPACK library not found in compilation");
+#else
+  char uplo = upper ? 'U' : 'L';
+
+  Tensor outTag =  out;
```
Comment (vishwakftw) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:581]: This assignment can be done in the else clause of the branch `batch_size > 1` (line 590).

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -571,6 +571,100 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
...
+    outTag =  out.flatten(0, -3);
+    selfTag = self.flatten(0, -3);
+  }
+  
+  scalar_t* self_data = selfTag.data_ptr<scalar_t>();
+  scalar_t* out_data = outTag.data_ptr<scalar_t>();
+  /* TODO: do this in parallel */
+  for (int64_t i = 0; i < batch_size; i++) {
+    scalar_t* out_working_ptr = &out_data[i * self_matrix_stride];
+    scalar_t* self_working_ptr = &self_data[i * self_matrix_stride];
```
Comment (vishwakftw) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:600]: `self_working_ptr` is not required.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 6]
```diff
@@ -571,6 +571,100 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
...
+  auto batch_size = batchCount(self);
+  auto n = self.size(-2);
+  auto m = self.size(-1);
+
+  if (batch_size > 1) {
+    outTag =  out.flatten(0, -3);
+    selfTag = self.flatten(0, -3);
+  }
+  
+  scalar_t* self_data = selfTag.data_ptr<scalar_t>();
```
Comment (vishwakftw) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:595]: `self_data` is not required.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu [hunk 7]
```diff
@@ -901,6 +901,95 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
...
+    "compilation. Please rebuild with MAGMA.");
+#else
+  magma_uplo_t uplo = upper ? MagmaUpper : MagmaLower;
+  Tensor outTag =  out;
+  Tensor selfTag = self;
+
+  auto self_matrix_stride = matrixStride(self);
+  auto batch_size = batchCount(self);
+  auto n = self.size(-2);
+  auto m = self.size(-1);
```
Comment (vishwakftw) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:917]: Same as above, the input is square.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu [hunk 8]
```diff
@@ -901,6 +901,95 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
...
+
+  if (batch_size > 1) {
+    outTag =  out.flatten(0, -3);
+    selfTag = self.flatten(0, -3);
+  }
+  
+  scalar_t* self_data = selfTag.data_ptr<scalar_t>();
```
Comment (vishwakftw) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:924]: Same as above: `self_data` is not required.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu [hunk 9]
```diff
@@ -901,6 +901,95 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
...
+  scalar_t* out_data = outTag.data_ptr<scalar_t>();
+  int info;
+  /* TODO: do this in parallel */
+  for (int64_t i = 0; i < batch_size; i++) {
+    scalar_t* out_working_ptr = &out_data[i * self_matrix_stride];
+    scalar_t* self_working_ptr = &self_data[i * self_matrix_stride];
```
Comment (vishwakftw) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:930]: Same as above: `self_working_ptr` is not required.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu [hunk 10]
```diff
@@ -901,6 +901,95 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
...
+    double err_val = 1e-8;
+    magma_int_t info = 0;
+    magmaCholesky<scalar_t>(uplo, n, out_working_ptr, n, &info);
+    for (int64_t j=0; j<8; j++)
+    {
+      if (info == 0) {
+        break;
+      }
+      // lapackChelesky failed.
```
Comment (vishwakftw) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:939]: nit: `magmaCholesky`.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu [hunk 11]
```diff
@@ -901,6 +901,95 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
...
+  if (batch_size > 1) {
+    outTag =  out.flatten(0, -3);
+    selfTag = self.flatten(0, -3);
+  }
+  
+  scalar_t* self_data = selfTag.data_ptr<scalar_t>();
+  scalar_t* out_data = outTag.data_ptr<scalar_t>();
+  int info;
+  /* TODO: do this in parallel */
+  for (int64_t i = 0; i < batch_size; i++) {
```
Comment (vishwakftw) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:928]: This is probably not ideal. What could be tried instead is: run `magmaCholeskyBatched` to get the list of infos pertaining to the batch, and then filter out the non-PD matrices, and add jitter to them alone. This method, according to me, is not leveraging MAGMA's parallelism capabilities.

File: aten/src/ATen/native/native_functions.yaml [hunk 12]
```diff
@@ -4524,6 +4524,17 @@
     CPU: _cholesky_helper_cpu
     CUDA: _cholesky_helper_cuda
 
+- func: cholesky_mod(Tensor self, bool upper=False) -> (Tensor result, Tensor err)
```
Comment (vishwakftw) [inline aten/src/ATen/native/native_functions.yaml:4527]: Perhaps something worth discussing. Should we make this a separate function, or add an argument `jitter` (boolean) to the original function, which then dispatches internally to either variant.

**Review by vishwakftw — COMMENTED** [review]
I have left some comments. CPU version looks ok, but I think we can improve the CUDA specific implementation.

**Comment by ezyang** [comment]
@vishwakftw can I leave this review entirely to you? I'll merge when you approve.

**Comment by vishwakftw** [comment]
Sure!

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (Balandat) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:601]: Yeah `err_val` is confusing, `jitter` is the standard term.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 13]
```diff
@@ -571,6 +571,100 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
...
+  for (int64_t i = 0; i < batch_size; i++) {
+    scalar_t* out_working_ptr = &out_data[i * self_matrix_stride];
+    scalar_t* self_working_ptr = &self_data[i * self_matrix_stride];
+    double err_val = 1e-8;
+    int info;
+    lapackCholesky<scalar_t>(uplo, n, out_working_ptr, n, &info);
+    for (int64_t j=0; j<8; j++)
+    {
+      if (info == 0) {
+        break;
```
Comment (Balandat) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:607]: Just to make sure: This circumvents the exceedingly slow error handling that is pointed out in #34272, right?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 14]
```diff
@@ -571,6 +571,100 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
...
+  auto m = self.size(-1);
+
+  if (batch_size > 1) {
+    outTag =  out.flatten(0, -3);
+    selfTag = self.flatten(0, -3);
+  }
+  
+  scalar_t* self_data = selfTag.data_ptr<scalar_t>();
+  scalar_t* out_data = outTag.data_ptr<scalar_t>();
+  /* TODO: do this in parallel */
```
Comment (Balandat) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:597]: this seems like an important TODO - does the current vanilla cholesky implementation do this in parallel?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 15]
```diff
@@ -571,6 +571,100 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
...
+      for (int64_t mm=0; mm<m; mm++)
+      {
+        // Add err_val to the diagonal
+        out_working_ptr[(mm * m) + mm] += err_val;
+      }
+      err_data[i] = err_val;
+      err_val *= 100.0;
+      lapackCholesky<scalar_t>(uplo, n, out_working_ptr, n, &info);
+    }
+    infos[i] = info;
```
Comment (Balandat) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:624]: it would be good to also return the actual jitter values that were used in the computation as debug information

File: aten/src/ATen/native/native_functions.yaml (hunk 12 above)
Comment (Balandat) [reply aten/src/ATen/native/native_functions.yaml:4527]: I think the jitter arg would make sense. But we should distinguish between a function that allows to specify a jitter value and the one here that automatically increases jitter until the decomposition succeeds.

**Comment by Balandat** [comment]
In general, I do believe that we will want to be somewhat more sophisticated on the jitter handling than just hard-coding increasing values. See https://pdfs.semanticscholar.org/74df/60c51d8f56ff514534ae545970ca7f456610.pdf and references therein for some more sophisticated approaches (though these probably would require going deeper than just calling lapack/magma under the hood. 

In particular, if the error handling issues from #34272 apply also in this PR, how much value does the current PR provide over some simple python implementation as in https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/utils/cholesky.py#L11 ?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 14 above)
Comment (vishwakftw) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:597]: No, the vanilla Cholesky doesn't do this in parallel.

File: aten/src/ATen/native/native_functions.yaml (hunk 12 above)
Comment (vishwakftw) [reply aten/src/ATen/native/native_functions.yaml:4527]: Could there be modes? For instance: None, 'auto', jitter_val, or something like that.

Comment (Balandat) [reply aten/src/ATen/native/native_functions.yaml:4527]: Yeah I think that's reasonable. The python interface could look sth like

```
def cholesky(A: Tensor: upper: bool=False, jitter: Optional[Union[str, float]]='none') -> ...
```
where `jitter='none'` is the current behavior, `jitter='auto'` is the behavior in this PR, and `jitter=value` uses the specified jitter value. This is quite a bit of arg overloading, but could be ok. The one thing this wouldn't allow is to have the behavior in this PR starting from a provided value.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 13 above)
Comment (mattip) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:607]: As long as the jitter addition succeeds (after maximum 6 iterations) we avoid the error handling. Otherwise it still raises. I should add a corresponding test

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 15 above)
Comment (mattip) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:624]: All of them? The successful value is returned in `e` (which I will rename to `jitter`)

**Comment by mattip** [comment]
> In general, I do believe that we will want to be somewhat more sophisticated on the jitter handling 

I think this could be done via the `jitter` argument to `cholesky` (as proposed elsewhere): 
- `jitter=None` (default) - no jitter iterations
- `jitter=[sequence of values]` - try the values in order

This would have the advantage of allowing exploration of different jitter strategies without requiring recompilation. Perhaps we could add a `jitter="auto"` in a future version once a good strategy has emerged. We could also allow `jitter=scalar` without requiring wrapping it as a sequnce `jitter=[scalar]`

**Comment by Balandat** [comment]
> I think this could be done via the jitter argument to cholesky (as proposed elsewhere):

So that would address the API aspects of this, but not the underlying algorithm. I'm not really familiar with the literature here, but it seems there are ways to do this in a much more sophisticated adaptive way rather than trying a sequence of increasing values. But I guess we need to start somewhere.

File: aten/src/ATen/native/native_functions.yaml (hunk 12 above)
Comment (mattip) [reply aten/src/ATen/native/native_functions.yaml:4527]: I am not sure how to do this in the `native_functions.yaml` interface definition. It seems the best I can do is an optional Tensor. We could write a python-only wrapper, but then the interface in the C++ code would differ.

**Comment by mattip** [comment]
I rewrote `cholesky_mod` to accept an optional  1d Tensor of double values (`jitter_in`). These are iterated over in order until the modified cholesky succeeds, and the successful value returned (`jitter_out`). In the "blocked" variant, `jitter_out` matches the block dimensions.

Still TODO: improve CUDA performance. I "resolved" the conversations for review items I fixed.

**Comment by mattip** [comment]
Rebased off master. 

I also updated the backward cholesky_mod, but I am not sure this is correct. The whole point of the cholesky_mod function is to enable Cholesky decomposition when the input is close to singular, so the function is very sensitive to small pertubations in values. Small pertubations is exactly what `gradcheck` does (for the numerical validation) so the check fails. I didn't commit code that uses close-to-singular matrices, I am still trying to find a way to create them. The current forward tests in `test_torch.py` simply zero out a few values on the diagonal, but that seems like too blunt a tool. 

So in summary:
- I still am looking for a way to generate proper cases to test the operator
- These cases should be used to test the backward operator
- My implementation of the backward operator when output jitter is non-zero may not be correct.

In addition:
- The CUDA implemenation still needs to be parallelized
- The original issue suggested some papers with algorithms for choosing the jitter values, these should be explored.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 14 above)
Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:607]: No need to do it in parallel so that all the threads are utilized in the kernels, which usually dominate the overall computation.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 13 above)
Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:616]: Raises you mean with the internal Torch checks?

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu (hunk 11 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:936]: Same as above, if on CPU, makes no sense to parallelize because of the underlying parallelization in Lapack, and if on CUDA, this serial host loop will enqueue the CUDA stream queues.

File: tools/autograd/derivatives.yaml [hunk 16]
```diff
@@ -297,6 +297,9 @@
 - name: cholesky(Tensor self, bool upper=False) -> Tensor
   self: cholesky_backward(grad, upper, result)
 
+- name: cholesky_mod(Tensor self, bool upper=False, *, Tensor? jitter=None) -> (Tensor result, Tensor jitter_out)
+  self: cholesky_mod_backward(grad, upper, result, jitter_out)
+
```
Comment (nikitaved) [inline tools/autograd/derivatives.yaml:302]: Maybe it makes sense to tag the gradient wrt to jitter as `non_differentiable`?

File: tools/autograd/templates/Functions.cpp [hunk 17]
```diff
@@ -852,6 +852,23 @@ Tensor cholesky_backward(Tensor grad, bool upper, Tensor L) {
...
+        std::cout << "jitter dim " << jitter.dim() << ", size[0] " << jitter.size(0) << std::endl;
+        L_tag.diagonal().add_(jitter[0]);
+    }
+    else {
+        L_tag.diagonal().add_(jitter);
+    }
+
+  }
+  return cholesky_backward(grad, upper, L_tag);
+}
```
Comment (nikitaved) [inline tools/autograd/templates/Functions.cpp:870]: Hmm, why `L_tag = L - I * jiter[0]`?

Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:870]: If we assume that `jitter` is a constant, then `A + e* I = L L^T` implies that `dA = dL L^T + L dL^T`, so the backward formula should be the same as for classic Cholesky.

File: tools/autograd/templates/Functions.cpp [hunk 18]
```diff
@@ -852,6 +852,23 @@ Tensor cholesky_backward(Tensor grad, bool upper, Tensor L) {
   return grad_input.add(grad_input.transpose(-1, -2)).mul_(0.5);  // Symmetrizing the gradient
 }
 
+Tensor cholesky_mod_backward(Tensor grad, bool upper, Tensor L, Tensor jitter) {
+  // L_tag = L - I * jiter[0], then use cholesky_backward
+  Tensor L_tag = L;
+  if (jitter.defined()) {
+    L_tag.copy_(L);
```
Comment (nikitaved) [inline tools/autograd/templates/Functions.cpp:859]:
```
Tensor  L_tag = L;
L_tag.copy_(L);
```
Could you please explain what this sequence is doing?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 13 above)
Comment (mattip) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:616]: Yes, with the checks of `info` beginning on line 646

File: tools/autograd/templates/Functions.cpp (hunk 17 above)
Comment (mattip) [reply tools/autograd/templates/Functions.cpp:870]: makes sense. I was making this too hard for no reason.

File: tools/autograd/derivatives.yaml (hunk 16 above)
Comment (mattip) [reply tools/autograd/derivatives.yaml:302]: right, makes sense.

File: tools/autograd/templates/Functions.cpp (hunk 18 above)
Comment (mattip) [reply tools/autograd/templates/Functions.cpp:859]: removed this function

**Comment by mruberry** [comment]
Hey @mattip, just checking in on this PR. Is it ready to be reviewed?

**Comment by mattip** [comment]
While a review is always nice, this still needs a parallel CUDA implementation to be considered for merge. Also there is the higher-level discussion about the value for jitter. Currently, the user is expected to supply some guesses. It would be nice to try to figure out a theoretical optimal value and use that instead.

**Comment by mattip** [comment]
Closing: too low priority to continue..

**Comment by rgommers** [comment]
> Closing: too low priority to continue..

For the record, that doesn't seem to be the case. I'll take the action to find the right person to take this over.

**Comment by Balandat** [comment]
At the very least it would be good to expose a raw version of the underlying low-level Cholesky call that doesn't use torch's error and returns the raw success flags. That way we can at least detect if matrices are not pd much faster. B/c of this error handling overhead there currently is no fast way of checking for positive definiteness of matrices (that I know of) in torch.
